### PR TITLE
Fixes

### LINKED ISSUE #6: Doesn't work with nested node_modules
I use pnpm, and it seems like `pnpx cjs2esm` tries to import 5to6-codemod directly from node_modules but fails.

Once I added 5to6-codemod to the project it worked.

### LINKED ISSUE #7: Doesn't support ChainExpression
We use ?. extensively and it fails to parse this :cry:

### PR DESCRIPTION
**Pull request by wmertens**
Fixes #7 with new jscodemod
Fixes #6 by proper module path lookup

Allows all files but filters node_modules (but I guess you might not like this?)

My use case is converting an existing project once, so including all files in the root dir

### CODE REVIEW

**Comment by wmertens** [comment]
that last commit doesn't have passing tests, the tests are too much testing the implementation instead of the result.

File: src/transformer.js [hunk 1]
```diff
@@ -54,74 +89,111 @@ const transform = (file, api, options) => {
...
-      if (
-        (importPath === '.' || importPath.startsWith('./')) &&
-        !replacement.startsWith('./')
-      ) {
-        replacement = `./${replacement}`;
-      }
+      const replacement = updatePath(item, base);
 
       // Replace the node with a new one on the AST.
+      // return j.importDeclaration.from({ ...item.value, source: j.literal(replacement) });
```
Comment (homer0) [inline src/transformer.js:118]: ?

File: src/transformer.js [hunk 2]
```diff
@@ -54,74 +89,111 @@ const transform = (file, api, options) => {
...
+        source: j.literal(replacement),
+      });
+    });
   // =================================================
   // Parse the modules modifications.
   // =================================================
   if (cjs2esm.modules.length) {
+    const modules = cjs2esm.modules.map((r) => ({
+      ...r,
+      regex: new RegExp(`^${r.name}(/|$)`),
```
Comment (homer0) [inline src/transformer.js:158]:
```suggestion
      regex: new RegExp(`^${r.name}(?:/|$)`),
```

No need to capture it.

File: src/transformer.js [hunk 3]
```diff
@@ -54,74 +89,111 @@ const transform = (file, api, options) => {
...
+    }));
+
     root
       .find(j.ImportDeclaration)
       // Filter out the import statments that don't need to be modified.
       .filter((item) => {
         const importPath = item.value.source.value;
-        return cjs2esm.modules.some((mod) => importPath.startsWith(mod.name));
+        return modules.some((mod) => mod.regex.test(importPath));
```
Comment (homer0) [inline src/transformer.js:166]:
```suggestion
        return modules.some((mod) => importPath.match(mod.regex));
```

Yes, I'm being picky here: even though you didn't use `g`, I only use `.test` when I need the `lastIndex`.

File: src/transformer.js [hunk 4]
```diff
@@ -54,74 +89,111 @@ const transform = (file, api, options) => {
...
+    //     const exportPath = item.value.source.value;
+    //     const info = cjs2esm.modules.find((mod) => exportPath.startsWith(mod.name));
+    //     const find = info.find ? new RegExp(info.find) : new RegExp(`^${info.name}`);
+    //     const replacement = exportPath.replace(find, info.path);
+
+    //     return j.exportNamedDeclaration.from({
+    //       ...item.value,
+    //       source: j.literal(replacement),
+    //     });
+    //   });
```
Comment (homer0) [inline src/transformer.js:196]: Can you give me the expected input and output of the transformation that you are looking for in here?

File: src/index.js [hunk 5]
```diff
@@ -114,7 +114,9 @@ const ensureOutput = async (output) => {
  */
 const findFiles = async (directory) => {
   let result = await fs.readdir(directory);
-  result = result.filter((item) => !item.startsWith('.'));
+  result = result.filter(
+    (item) => !(item === '.' || item === '..' || item === 'node_modules'),
```
Comment (homer0) [inline src/index.js:118]:
```suggestion
    (item) => !['.', '..', 'node_modules'].includes(item),
```

File: tests/index.test.js [hunk 6]
```diff
@@ -300,7 +309,7 @@ describe('index', () => {
       expect(fs.readdir).toHaveBeenNthCalledWith(1, src);
       expect(fs.readdir).toHaveBeenNthCalledWith(2, path.join(src, 'utils'));
       expect(fs.stat).toHaveBeenCalledTimes(
-        filesRoot.length + filesUtils.length - 1, // .eslintrc
+        filesRoot.length + filesUtils.length - 3, // . .. node_modules
```
Comment (homer0) [inline tests/index.test.js:312]: You added `.`, `..`, `node_modules` and an extra `isDirectory: () => false`, but `findFiles` filters them before calling `fs.stat`: `!['.', '..', 'node_modules'].includes(item)`, so, why extra mock, and the change in here?

**Review by homer0 — COMMENTED** [review]
Hey 👋!

Sorry, I never got the notifications for the issues nor the PR (I thought I had the settings fixed 😞), but I saw a fork icon on my GH profile and then I found them 😅.

This looks good overall, but there are a few details:

- Check the jscodeshift upgrade.
- Can you point this to the `next` branch? I was actually planning on dropping Node 12 and Yarn (as default) soon, and I prefer to do a big breaking release.
- Yes, the tests go too much into the implementation, but I use them mostly as doc on how the parsing works and what do I expect on each step, rather than adding logs/breakpoints every time I need to try something or make a change... AST updates is still something hard for me :P.

File: package.json [hunk 7]
```diff
@@ -20,7 +20,7 @@
     "chalk": "^4.1.2",
     "del": "^6.0.0",
     "fs-extra": "^10.0.0",
-    "jscodeshift": "0.10.0"
+    "jscodeshift": "0.13.1"
```
[the remark this answers is not in the document]
Comment (homer0) [reply package.json:23]: Solved in #15

**Pull request closed without merging**